contracts/Marketplace.sol

```diff
@@ -1,74 +1,296 @@
+
+  function getServiceIndex(bytes memory sid) public view returns (uint serviceIndex) {
+    for (uint i = 0; i < services.length; i++) {
+      if (compareBytes(services[i].sid, sid)) {
+        return i;
+      }
     }
+    require(false, "Service not found");
+  }
 
-    function hasAccessToVersion(bytes32 version, address user) public view returns (bool) {
-        return _versionAccess[version][user];
+  function getServiceVersionIndex(uint serviceIndex, bytes20 hash) public view returns (uint versionIndex) {
+    Service storage service = services[serviceIndex];
+    for (uint i = 0; i < service.versions.length; i++) {
+      if (service.versions[i].hash == hash) {
+        return i;
+      }
     }
-}
+    require(false, "Version not found");
+  }
 
-contract Marketplace is Ownable, Pausable {
-    using Address for address;
+  function getServicePaymentIndex(uint serviceIndex, address purchaser) public view returns (uint paymentIndex) {
+    Service storage service = services[serviceIndex];
+    for (uint i = 0; i < service.payments.length; i++) {
+      if (service.payments[i].purchaser == purchaser) {
+        return i;
+      }
+    }
+    require(false, "Payment not found");
+  }
 
-    mapping (bytes32 => address) public serviceContracts;
-    bytes32[] public services;
-    uint256 public totalServices;
-    
-    event ServiceCreated(bytes32 sid, address serviceAddress);
+  // Count
 
-    constructor() Ownable() Pausable() public {}
+  function getServicesCount() public view returns (uint servicesCount) {
+    return services.length;
+  }
 
-    function createService(bytes32 sid) public whenNotPaused returns (address) {
-        if (serviceContracts[sid].isContract()) {
-            return;
-        }
-        serviceContracts[sid] = new Service(msg.sender, sid);
-        services.push(sid);
-        totalServices = totalServices + 1;
-        emit ServiceCreated(sid, serviceContracts[sid]);
-        return serviceContracts[sid];
+  function getServiceVersionsCount(uint serviceIndex) public view returns (uint serviceVersionsCount) {
+    return services[serviceIndex].versions.length;
+  }
+
+  function getServicePaymentsCount(uint serviceIndex) public view returns (uint servicePaymentsCount) {
+    return services[serviceIndex].payments.length;
+  }
+
+  // ------------------------------------------------------
+  // Setter functions
+  // ------------------------------------------------------
+
+  // Manage Service
+
+  function createService (bytes memory sid, uint256 price) public whenNotPaused returns (uint serviceIndex) {
+    require(!isServiceSidExist(sid), "Service's sid is already used");
+    services.length++;
+    Service storage service = services[services.length - 1];
+    service.sid = sid;
+    service.price = price;
+    service.owner = msg.sender;
+    emit ServiceCreated(
+      services.length - 1,
+      service.sid,
+      service.owner,
+      service.price
+    );
+    return services.length - 1;
```

Comment:
is it return needed here? in order to execute this function user need to call send() with some gas limit and price and this function always return hash transaction.

Reply:
In ERC20 contract from open zeppelin, they always return a bool.
What about doing like them?

Reply:
we can, but what's the point, why they return true, we need to figure it out :)